### DISCUSSION TITLE
Support for uvloop in socketio.AsyncClient()

### DISCUSSION THREAD

**Discussion by samiamlabs** [miguelgrinberg/python-socketio#759] [Q&A] [answered]
Hi!

I'm trying to use python-socketio for signalling when establishing a WebRTC connection with the aiortc library.
I could really use the performance boost uvloop for processing WebRCT frames in the main asyncio thread,
but it does not seem to work with socketio. I can't find anything about `AsyncClient` having support for uvloop so I assume it is not supported?

Would adding uvloop support to `AsyncClient` be difficult? Was thinking that I could maybe attempt to implement this myself on a fork if the current developers don't have time. I can probably find another solution to my problem if it's a hard problem to solve though...

**Comment by miguelgrinberg** [comment, marked as answer]
@samiamlabs I don't believe I have done anything in this package that prevents you from using uvloop. I did not intend to, at least. The asyncio client should happily use the loop that you set. Have you tried?

**Reply by samiamlabs** [reply]
Ok! 
I have tried using something like:

```python
sio = socketio.AsyncClient()

    @sio.event
    def connect():
        print("Connected to the signaling server!")

    @sio.event
    def connect_error(data):
        print("The connection failed!")

    @sio.event
    def disconnect():
        print("I'm disconnected!")

    async def handle_connection_offer(data):
        print('Recieved WebRTC connection offer')
        conn = ConnectionHandler(node)
        server_response = await conn.get_local_description(data['offer'])

        return server_response

    async def handle_connection_offer_third_party(data):
        print('Recieved WebRTC connection offer with third party ice servers')
        conn = ConnectionHandler(node, use_third_party_ice_servers=True)
        server_response = await conn.get_local_description(data['offer'])

        return server_response

    async def connect_websocket():
        while(not sio.connected):
            try: 
                await sio.connect(CONNECTION_SERVER_HOSTNAME)
            except socketio.exceptions.ConnectionError:
                print("Conneciton error, trying again in 2 seconds!")
            except asyncio.TimeoutError:
                print("Timeout when trying to connect, trying again in 2 seconds!")

            await asyncio.sleep(2.0)

        sio.on('webrtc_connection_offer', handle_connection_offer)
        sio.on('webrtc_connection_offer_third_party', handle_connection_offer_third_party)

    async def cleanup():
        await ConnectionHandler.cleanup() # Close all connections
        await sio.disconnect()

    asyncio.ensure_future(connect_websocket())

    asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
    loop = asyncio.get_event_loop()
    try:
        loop.run_forever()
    except KeyboardInterrupt:
        pass
    finally:
        loop.set_debug(True)
        loop.run_until_complete(cleanup())
        print('Closed all connections, shutting down')

```

This seems to connect to the signalling server with the normal loop but nothing happens with uvloop.
Could be something else in my code that is going wrong though. 

Have not tried it in a minimal example and it is possible that it works. Should probably have done that to make sure before opening this issue... sorry about that.

**Reply by miguelgrinberg** [reply]
You haven't included the code where the loop is created. Are you sure you are installing uvloop as your default loop before you do anything related to asyncio? The code above seems to suggest you start the connect_websocket function in some loop before you set the loop policy. That seems wrong to me.

**Reply by samiamlabs** [reply]
Haha, I'm really not very experienced with asyncio so it is likely that I'm doing all kinds of things wrong...

Yup, you were right. Needed to set the loop policy before `asyncio.ensure_future(connect_websocket())`.
Works now, thanks a lot for the help! 

I should not be using issues as a support forum though... sorry about that, will try to do better in the future :)
